File: skills/webhook-consumer-architecture/examples/go/net-http/authentication.go
package examples

import (
	"errors"
	"net/http"
)

type WebhookAuthenticator interface {
	Authenticate(r *http.Request) error
}

type BasicAuthWebhookAuthenticator struct {
	ExpectedUsername string
	ExpectedPassword string
}

func (a BasicAuthWebhookAuthenticator) Authenticate(r *http.Request) error {
	username, password, ok := r.BasicAuth()
	if !ok {
		return errors.New("missing basic auth credentials")
	}

	if username != a.ExpectedUsername || password != a.ExpectedPassword {
		return errors.New("invalid basic auth credentials")
	}

	return nil
}

File: skills/webhook-consumer-architecture/examples/python/django/authentication.py
import base64

from rest_framework import exceptions
from rest_framework.authentication import BaseAuthentication


class WebhookBasicAuthentication(BaseAuthentication):
    expected_username = None
    expected_password = None

    def authenticate(self, request):
        auth_header = request.headers.get("Authorization")
        if not auth_header or not auth_header.startswith("Basic "):
            raise exceptions.AuthenticationFailed("Invalid basic authentication credentials")

        encoded = auth_header.split(" ", 1)[1]

        try:
            decoded = base64.b64decode(encoded).decode("utf-8")
            username, password = decoded.split(":", 1)
        except (TypeError, ValueError, UnicodeDecodeError):
            raise exceptions.AuthenticationFailed("Invalid basic authentication credentials")

        if username != self.expected_username or password != self.expected_password:
            raise exceptions.AuthenticationFailed("Invalid username or password")

        return None, None
